src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Utf8StringMarshaller.cs

```diff
@@ -0,0 +1,103 @@
+        // maximum number of bytes per 'char' is 3.
+        private const int MaxByteCountPerChar = 3;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Utf8StringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        public Utf8StringMarshaller(string str)
+            : this(str, default(Span<byte>))
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Utf8StringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        /// <param name="buffer">Buffer that may be used for marshalling.</param>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerFeatures.CallerAllocatedBuffer"/>
+        /// </remarks>
+        public Utf8StringMarshaller(string str, Span<byte> buffer)
+        {
+            allocated = null;
+            span = default;
+            if (str is null)
+                return;
+
+            // + 1 for number of characters in case left over high surrogate is ?
+            // * <MaxByteCountPerChar> (3 for UTF-8)
+            // +1 for null terminator
+            if (buffer.Length >= (str.Length + 1) * MaxByteCountPerChar + 1)
+            {
+                int byteCount = Encoding.UTF8.GetBytes(str, buffer);
+                buffer[byteCount] = 0; // null-terminate
+                span = buffer;
+            }
+            else
+            {
+                allocated = (byte*)Marshal.StringToCoTaskMemUTF8(str);
+            }
+        }
+
+        /// <summary>
+        /// Returns the native value representing the string.
+        /// </summary>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerFeatures.TwoStageMarshalling"/>
+        /// </remarks>
+        public byte* ToNativeValue() => allocated != null ? allocated : (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(span));
+
+        /// <summary>
+        /// Sets the native value representing the string.
+        /// </summary>
+        /// <param name="value">The native value.</param>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerFeatures.TwoStageMarshalling"/>
+        /// </remarks>
+        public void FromNativeValue(byte* value) => allocated = value;
+
+        /// <summary>
+        /// Returns the managed string.
+        /// </summary>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerDirection.Out"/>
+        /// </remarks>
+        public string? ToManaged() => allocated == null ? null : Marshal.PtrToStringUTF8((IntPtr)allocated);
+
+        /// <summary>
+        /// Frees native resources.
+        /// </summary>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerFeatures.UnmanagedResources"/>
+        /// </remarks>
+        public void FreeNative()
+        {
+            if (allocated != null)
```

Comment:
Both `PtrToStringUTF8` and `FreeCoTaskMem` have an early out with a null check. It should not be necessary to duplicate the null checks here.

Reply:
I guess it may be worth it to duplicate the null check for a tiny bit better perf if we believe that we are going to end up here with null most of the time.